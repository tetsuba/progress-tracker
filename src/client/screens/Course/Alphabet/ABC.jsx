import React, { useContext, useEffect } from 'react'
import { useParams } from 'react-router-dom'
import { useQuery } from '@apollo/react-hooks'
import { Button, Col, Container, ListGroup, Row } from 'react-bootstrap'

// COMPONENTS
import BreadCrumbs from '../../../components/BreadCrumbs/BreadCrumbs'
import Alphabet from '../../../components/Alphabet/Alphabet'
import { CRUMBS_KEY } from '../../../components/BreadCrumbs/crumbs'
import Histogram from '../../../components/Chart/Histogram/Histogram'
import AlphabetModal from '../../../components/Modal/AlphabetModal'

// CONTEXT
import { ModalContext } from '../../../context/ModalContext'

// ICONS
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faPlus } from '@fortawesome/free-solid-svg-icons'

// QUERIES
import { GET_STUDENT_COURSE_ABC } from '../../../api/course/courseABC/courseABC.query'
import { formatDataForHistogram } from './alphabet-utils'

export default function ABC() {
  const { id } = useParams()
  const variables = {
    input: {
      id: id || '',
    },
  }
  const options = {
    variables,
  }
  // $FlowFixMe - Investigate how to fix this flow issue
  const { loading, data } = useQuery(GET_STUDENT_COURSE_ABC, options)
  const { toggleModal, addTemplate } = useContext(ModalContext)

  useEffect(() => {
    addTemplate(AlphabetModal, { id: id, data: data })
  }, [addTemplate, data, id])

  if (loading) return <div>LOADING</div>

  return (
    <Container>
      <Row className="mt-5">
        <BreadCrumbs
          crumbKey={CRUMBS_KEY.ABC}
          name={data.getStudentCourseABC.studentName}
          id={id}
        />
      </Row>
      <Row className="mt-5">
        <h1>Alphabet (lowercase) assessment</h1>
      </Row>

      <Button
        className="float-right"
        variant="primary"
        onClick={() => toggleModal()}
      >
        <FontAwesomeIcon icon={faPlus} /> Add assessment
      </Button>
      <Row className="mt-5">
        <Col>
          <Histogram data={formatDataForHistogram(data)} />
        </Col>
      </Row>
      <Row className="mt-5">
        <Col>
          <h3>History:</h3>
          <ListGroup variant="flush">
            {data.getStudentCourseABC.history
              .map(({ date, alphabet }) => (
                <ListGroup.Item key={date}>
                  <div>{new Date(Number(date)).toDateString()}</div>
                  <Alphabet letters={alphabet} />
                </ListGroup.Item>
              ))
              .reverse()}
          </ListGroup>
        </Col>
      </Row>
    </Container>
  )
}
