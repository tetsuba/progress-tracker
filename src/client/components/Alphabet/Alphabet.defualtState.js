import type { AlphabetType } from './Alphabet.type'

const ALPHABET_DEFAULT_STATE: Array<AlphabetType> = [
  { letter: 'A', value: 0 },
  { letter: 'B', value: 0 },
  { letter: 'C', value: 0 },
  { letter: 'D', value: 0 },
  { letter: 'E', value: 0 },
  { letter: 'F', value: 0 },
  { letter: 'G', value: 0 },
  { letter: 'H', value: 0 },
  { letter: 'I', value: 0 },
  { letter: 'J', value: 0 },
  { letter: 'K', value: 0 },
  { letter: 'L', value: 0 },
  { letter: 'M', value: 0 },
  { letter: 'N', value: 0 },
  { letter: 'O', value: 0 },
  { letter: 'P', value: 0 },
  { letter: 'Q', value: 0 },
  { letter: 'R', value: 0 },
  { letter: 'S', value: 0 },
  { letter: 'T', value: 0 },
  { letter: 'U', value: 0 },
  { letter: 'V', value: 0 },
  { letter: 'W', value: 0 },
  { letter: 'X', value: 0 },
  { letter: 'Y', value: 0 },
  { letter: 'Z', value: 0 },
]

export default ALPHABET_DEFAULT_STATE
