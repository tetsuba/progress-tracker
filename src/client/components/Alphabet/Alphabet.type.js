export type AlphabetType = {
  letter: string,
  value: number,
}
