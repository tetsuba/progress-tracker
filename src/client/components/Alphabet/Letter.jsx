import React from 'react'

// TYPES
import type { AlphabetType } from './Alphabet.type'

type Props = {
  update?: ((state: Array<AlphabetType>) => Array<AlphabetType>) => void,
  key?: ?string,
}

export default function Letter(props: Props & AlphabetType) {
  const { letter, value, update } = props
  let colour
  let bgColour
  switch (value) {
    case 1:
      bgColour = '#309c3f'
      colour = '#FFF'
      break
    case 2:
      bgColour = '#e33f3f'
      colour = '#FFF'
      break
    default:
      bgColour = '#fff'
      colour = '#000'
  }

  const styles = {
    width: !update ? 'auto' : '80px',
    height: !update ? 'auto' : '100px',
    display: 'flex',
    justifyContent: 'center',
    alignItems: 'center',
    border: 'solid 1px #000',
    fontSize: !update ? '14px' : '42px',
    margin: '4px',
    color: colour,
    backgroundColor: bgColour,
    padding: !update ? '8px' : '',
    cursor: 'pointer',
  }

  function updateLetter(state: Array<AlphabetType>): Array<AlphabetType> {
    function updateValue(value: number) {
      return value >= 2 ? 0 : (value += 1)
    }

    return state.map((obj) => ({
      ...obj,
      value: obj.letter === letter ? updateValue(obj.value) : obj.value,
    }))
  }

  return (
    <div style={styles} onClick={() => update && update(updateLetter)}>
      {letter}
    </div>
  )
}
