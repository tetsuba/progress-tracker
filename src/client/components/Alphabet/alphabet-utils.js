// TODO: update types
type Props = {
  getAlphabetAssessment: {
    history: [any],
  },
}

export function formatDefaultState(data: Props) {
  if (data.getAlphabetAssessment.history.length < 1) {
    return undefined
  }
  const eof = data.getAlphabetAssessment.history.length - 1
  const state = data.getAlphabetAssessment.history[eof].alphabet
  return state.map(({ letter, value }) => ({ letter, value }))
}
