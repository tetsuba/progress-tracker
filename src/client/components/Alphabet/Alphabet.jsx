import React from 'react'

// COMPONENTS
import Letter from './Letter'

// TYPES
import type { AlphabetType } from './Alphabet.type'

type Props = {
  letters: Array<{ letter: string, value: number }>,
  update?: ((state: Array<AlphabetType>) => Array<AlphabetType>) => void,
}

export default function Alphabet(props: Props) {
  const { letters, update } = props
  return (
    <div style={{ display: 'flex', alignItems: 'left', flexWrap: 'wrap' }}>
      {letters.map((props) => (
        <Letter {...props} update={update} />
      ))}
    </div>
  )
}
