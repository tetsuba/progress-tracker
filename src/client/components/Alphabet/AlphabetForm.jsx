import React, { useEffect, useState } from 'react'
import { useMutation } from '@apollo/react-hooks'
import { Button } from 'react-bootstrap'

// CONTEXT
import { LoadingContext } from '../../context/LoadingContext'

// COMPONENTS
import Alphabet from './Alphabet'

// QUERIES
import { GET_ALPHABET_ASSESSMENT } from '../../api/assessment/alphabet/alphabet.query'

// MUTATIONS
import { SAVE_ALPHABET_ASSESSMENT_MUTATION } from '../../api/assessment/alphabet/alphabet.mutaion'

// DEFAULT STATE
import ALPHABET_DEFAULT_STATE from './Alphabet.defualtState'
import { formatDefaultState } from './alphabet-utils'

// TODO: add loading indicator
type Props = {
  id: string,
  toggleModal: () => void,
  data: any,
  type: string,
}

export default function AlphabetForm(props: Props) {
  const { toggleModal, id, data, type } = props
  const defaultState =
    formatDefaultState(data) ||
    (type === 'lowerCase'
      ? ALPHABET_DEFAULT_STATE.map((state) => ({
          ...state,
          letter: state.letter.toLocaleLowerCase(),
        }))
      : ALPHABET_DEFAULT_STATE)

  const variables = {
    input: {
      studentId: id || '',
      typeOfAlphabetAssessment: type || '',
    },
  }

  const { showLoading, hideLoading } = React.useContext(LoadingContext)

  const [letters, setLetters] = useState(defaultState)
  const [saveAlphabetAssessment] = useMutation(
    SAVE_ALPHABET_ASSESSMENT_MUTATION,
    {
      refetchQueries: [{ query: GET_ALPHABET_ASSESSMENT, variables }],
    }
  )

  useEffect(() => {
    setLetters(defaultState)
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [])

  return (
    <>
      <Alphabet letters={letters} update={setLetters} />
      <Button
        className="float-right"
        onClick={() => {
          const variables = {
            input: {
              studentId: id,
              date: String(Date.now()),
              alphabet: letters,
              typeOfAlphabetAssessment: type,
            },
          }
          showLoading()
          saveAlphabetAssessment({ variables })
            .then(() => {
              setLetters(defaultState)
              toggleModal()
              hideLoading()
            })
            .catch(() => {
              // TODO: An error message should be displayed to the user
              toggleModal()
              hideLoading()
            })
        }}
      >
        Save Progress
      </Button>
    </>
  )
}
