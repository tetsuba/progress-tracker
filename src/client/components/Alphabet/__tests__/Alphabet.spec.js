import { testRenderer } from '../../../../test/testHelper'
import Alphabet from '../../Alphabet/Alphabet'

describe('<Alphabet>', () => {
  const lettersMock = 'abcdefghijklnmopqrstuvwxyz'.split('').map((letter) => ({
    letter: letter,
    value: 0,
  }))

  describe('@Render', () => {
    it('Should render alphabet', () => {
      const props = {
        letters: lettersMock,
        update: jest.fn(),
      }
      const wrapper = testRenderer(Alphabet, props)
      expect(wrapper).toMatchSnapshot()
    })
  })
  describe('@Event', () => {
    describe('@onClick', () => {
      it('should trigger an update prop', () => {
        const props = {
          letters: lettersMock,
          update: jest.fn(),
        }
        const wrapper = testRenderer(Alphabet, props)
        wrapper.find('Letter').get(0).props.update()
        expect(props.update).toHaveBeenCalledTimes(1)
      })
    })
  })
})
