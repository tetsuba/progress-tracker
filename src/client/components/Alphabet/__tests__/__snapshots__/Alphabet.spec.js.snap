// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Alphabet> @Render Should render alphabet 1`] = `
<div
  style={
    Object {
      "alignItems": "left",
      "display": "flex",
      "flexWrap": "wrap",
    }
  }
>
  <Letter
    key="a-letter"
    letter="a"
    update={[MockFunction]}
    value={0}
  />
  <Letter
    key="b-letter"
    letter="b"
    update={[MockFunction]}
    value={0}
  />
  <Letter
    key="c-letter"
    letter="c"
    update={[MockFunction]}
    value={0}
  />
  <Letter
    key="d-letter"
    letter="d"
    update={[MockFunction]}
    value={0}
  />
  <Letter
    key="e-letter"
    letter="e"
    update={[MockFunction]}
    value={0}
  />
  <Letter
    key="f-letter"
    letter="f"
    update={[MockFunction]}
    value={0}
  />
  <Letter
    key="g-letter"
    letter="g"
    update={[MockFunction]}
    value={0}
  />
  <Letter
    key="h-letter"
    letter="h"
    update={[MockFunction]}
    value={0}
  />
  <Letter
    key="i-letter"
    letter="i"
    update={[MockFunction]}
    value={0}
  />
  <Letter
    key="j-letter"
    letter="j"
    update={[MockFunction]}
    value={0}
  />
  <Letter
    key="k-letter"
    letter="k"
    update={[MockFunction]}
    value={0}
  />
  <Letter
    key="l-letter"
    letter="l"
    update={[MockFunction]}
    value={0}
  />
  <Letter
    key="n-letter"
    letter="n"
    update={[MockFunction]}
    value={0}
  />
  <Letter
    key="m-letter"
    letter="m"
    update={[MockFunction]}
    value={0}
  />
  <Letter
    key="o-letter"
    letter="o"
    update={[MockFunction]}
    value={0}
  />
  <Letter
    key="p-letter"
    letter="p"
    update={[MockFunction]}
    value={0}
  />
  <Letter
    key="q-letter"
    letter="q"
    update={[MockFunction]}
    value={0}
  />
  <Letter
    key="r-letter"
    letter="r"
    update={[MockFunction]}
    value={0}
  />
  <Letter
    key="s-letter"
    letter="s"
    update={[MockFunction]}
    value={0}
  />
  <Letter
    key="t-letter"
    letter="t"
    update={[MockFunction]}
    value={0}
  />
  <Letter
    key="u-letter"
    letter="u"
    update={[MockFunction]}
    value={0}
  />
  <Letter
    key="v-letter"
    letter="v"
    update={[MockFunction]}
    value={0}
  />
  <Letter
    key="w-letter"
    letter="w"
    update={[MockFunction]}
    value={0}
  />
  <Letter
    key="x-letter"
    letter="x"
    update={[MockFunction]}
    value={0}
  />
  <Letter
    key="y-letter"
    letter="y"
    update={[MockFunction]}
    value={0}
  />
  <Letter
    key="z-letter"
    letter="z"
    update={[MockFunction]}
    value={0}
  />
</div>
`;
